### PR TITLE
add search topics

### PR DESCRIPTION
**Pull request by lucmski**

### CODE REVIEW

File: github/search.go [hunk 1]
```diff
@@ -71,6 +71,22 @@ func (s *SearchService) Repositories(ctx context.Context, query string, opt *Sea
 	return result, resp, err
 }
 
+// TopicsSearchResult represents the result of a topics search.
+type TopicsSearchResult struct {
+	Total             *int         `json:"total_count,omitempty"`
+	IncompleteResults *bool        `json:"incomplete_results,omitempty"`
+	Repositories      []Repository `json:"items,omitempty"`
```
Comment (gmlewis) [inline github/search.go]: I don't think these results are `Repositories`.

In fact, to match the other `Search` endpoints, you will most likely need to create a new struct called `TopicResult` and then fill in the fields and tags from the GitHub v3 API developer docs.

So this should be something like:

```go
Topics []*TopicResult `json:items,omitempty`
```

File: go.mod [hunk 2]
```diff
@@ -3,6 +3,8 @@ module github.com/google/go-github/v28
 require (
 	github.com/golang/protobuf v1.2.0 // indirect
 	github.com/google/go-querystring v1.0.0
+	github.com/k0kubun/pp v3.0.1+incompatible // indirect
```
Comment (gmlewis) [inline go.mod]: Please revert the changes to this file.

File: go.sum [hunk 3]
```diff
@@ -2,6 +2,12 @@ github.com/golang/protobuf v1.2.0 h1:P3YflyNX/ehuJFLhxviNdFxQPkGK5cDcApsge1SqnvM
 github.com/golang/protobuf v1.2.0/go.mod h1:6lQm79b+lXiMfvg/cZm0SGofjICqVBUtrP5yJMmIC1U=
 github.com/google/go-querystring v1.0.0 h1:Xkwi/a1rcvNg1PPYe5vI8GbeBY/jrVuDX5ASuANWTrk=
 github.com/google/go-querystring v1.0.0/go.mod h1:odCYkC5MyYFN7vkCjXpyrEuKhc/BUO6wN/zVPAxq5ck=
+github.com/k0kubun/pp v3.0.1+incompatible h1:3tqvf7QgUnZ5tXO6pNAZlrvHgl6DvifjDrd9g2S9Z40=
```
Comment (gmlewis) [inline go.sum]: Please revert the changes to this file.

File: example/topics/main.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
+// Copyright 2017 The go-github AUTHORS. All rights reserved.
```
Comment (gmlewis) [inline example/topics/main.go]: Please change `2017` to `2019`. We place the current year in each file as it is created. Thanks.

File: github/search.go [hunk 5]
```diff
@@ -71,6 +71,34 @@ func (s *SearchService) Repositories(ctx context.Context, query string, opt *Sea
...
+type TopicResult struct {
+	Name             *string  `json:"name,omitempty"`
+	DisplayName      *string  `json:"display_name,omitempty"`
+	ShortDescription *string  `json:"short_description,omitempty"`
+	Description      *string  `json:"description,omitempty"`
+	CreatedBy        *string  `json:"created_by,omitempty"`
+	CreatedAt        *string  `json:"created_at,omitempty"`
+	UpdatedAt        *string  `json:"updated_at,omitempty"`
+	Featured         *bool    `json:"featured,omitempty"`
+	Score            *float32 `json:"score,omitempty"`
```
Comment (gmlewis) [inline github/search.go]: Let's please use `float64` here for consistency with the other endpoints.

File: github/search.go [hunk 6]
```diff
@@ -71,6 +71,34 @@ func (s *SearchService) Repositories(ctx context.Context, query string, opt *Sea
...
+
+type TopicResult struct {
+	Name             *string  `json:"name,omitempty"`
+	DisplayName      *string  `json:"display_name,omitempty"`
+	ShortDescription *string  `json:"short_description,omitempty"`
+	Description      *string  `json:"description,omitempty"`
+	CreatedBy        *string  `json:"created_by,omitempty"`
+	CreatedAt        *string  `json:"created_at,omitempty"`
+	UpdatedAt        *string  `json:"updated_at,omitempty"`
+	Featured         *bool    `json:"featured,omitempty"`
```
Comment (gmlewis) [inline github/search.go]: Please add the `curated` field:

```go
Curated *bool `json:"curated,omitempty"`
```

File: github/search.go [hunk 7]
```diff
@@ -71,6 +71,34 @@ func (s *SearchService) Repositories(ctx context.Context, query string, opt *Sea
...
+	Topics            []*TopicResult `json:"items,omitempty"`
+}
+
+type TopicResult struct {
+	Name             *string  `json:"name,omitempty"`
+	DisplayName      *string  `json:"display_name,omitempty"`
+	ShortDescription *string  `json:"short_description,omitempty"`
+	Description      *string  `json:"description,omitempty"`
+	CreatedBy        *string  `json:"created_by,omitempty"`
+	CreatedAt        *string  `json:"created_at,omitempty"`
```
Comment (gmlewis) [inline github/search.go]: Please change `*string` to `*Timestamp` here and on the line below as well.

File: github/search.go [hunk 8]
```diff
@@ -71,6 +71,34 @@ func (s *SearchService) Repositories(ctx context.Context, query string, opt *Sea
...
+	UpdatedAt        *string  `json:"updated_at,omitempty"`
+	Featured         *bool    `json:"featured,omitempty"`
+	Score            *float32 `json:"score,omitempty"`
+}
+
+// Topics searches repositories via various topics.
```
Comment (gmlewis) [inline github/search.go]: This sentence is a little bit awkward and the query parameter can contain more than just topics, it can also include qualifiers. So how about if we make this read a bit more closely to the GitHub API docs and say something like:

```go
// Topics finds topics via various criteria. Results are sorted by best match.
// Please see https://help.github.com/en/articles/searching-topics for more
// information about search qualifiers.
```

File: go.mod [hunk 9]
```diff
@@ -3,6 +3,7 @@ module github.com/google/go-github/v28
 require (
 	github.com/golang/protobuf v1.2.0 // indirect
 	github.com/google/go-querystring v1.0.0
+	github.com/mattn/go-colorable v0.1.2 // indirect
```
Comment (gmlewis) [inline go.mod]: Please revert the contents of this file.

File: go.sum [hunk 10]
```diff
@@ -2,6 +2,10 @@ github.com/golang/protobuf v1.2.0 h1:P3YflyNX/ehuJFLhxviNdFxQPkGK5cDcApsge1SqnvM
 github.com/golang/protobuf v1.2.0/go.mod h1:6lQm79b+lXiMfvg/cZm0SGofjICqVBUtrP5yJMmIC1U=
 github.com/google/go-querystring v1.0.0 h1:Xkwi/a1rcvNg1PPYe5vI8GbeBY/jrVuDX5ASuANWTrk=
 github.com/google/go-querystring v1.0.0/go.mod h1:odCYkC5MyYFN7vkCjXpyrEuKhc/BUO6wN/zVPAxq5ck=
+github.com/mattn/go-colorable v0.1.2 h1:/bC9yWikZXAL9uJdulbSfyVNIR3n3trXl+v8+1sx8mU=
```
Comment (gmlewis) [inline go.sum]: Please revert the contents of this file.

Comment (lucmski) [reply go.sum]: done

File: github/search.go (hunk 8 above)
Comment (lucmski) [reply github/search.go]: done

File: github/search.go (hunk 7 above)
Comment (lucmski) [reply github/search.go]: done

File: github/search.go (hunk 6 above)
Comment (lucmski) [reply github/search.go]: done

File: github/search.go (hunk 5 above)
Comment (lucmski) [reply github/search.go]: done

File: example/topics/main.go (hunk 4 above)
Comment (lucmski) [reply example/topics/main.go]: done

File: go.mod (hunk 9 above)
Comment (lucmski) [reply go.mod]: done

File: github/search.go (hunk 7 above)
Comment (lucmski) [reply github/search.go]: # github.com/google/go-github/v28/github
../../github/github-accessors.go:11924:9: cannot use *t.CreatedAt (type Timestamp) as type string in return argument
../../github/github-accessors.go:11972:10: cannot use t.Score (type *float64) as type *float32 in return argument

Comment (gmlewis) [reply github/search.go]: You need to re-run `go generate ./...`.  Please see `CONTRIBUTING.md` for details.

**Comment by gmlewis** [comment]
Thank you, @wesleimp!
Merging.
